Mathlib/GroupTheory/Index.lean

```diff
@@ -0,0 +1,622 @@
+/-
+Copyright (c) 2021 Thomas Browning. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning
+
+! This file was ported from Lean 3 source module group_theory.index
+! leanprover-community/mathlib commit dc6c365e751e34d100e80fe6e314c3c3e0fd2988
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Finite.Card
+import Mathlib.GroupTheory.Finiteness
+import Mathlib.GroupTheory.GroupAction.Quotient
+
+/-!
+# Index of a Subgroup
+
+In this file we define the index of a subgroup, and prove several divisibility properties.
+Several theorems proved in this file are known as Lagrange's theorem.
+
+## Main definitions
+
+- `H.index` : the index of `H : Subgroup G` as a natural number,
+  and returns 0 if the index is infinite.
+- `H.relindex K` : the relative index of `H : Subgroup G` in `K : Subgroup G` as a natural number,
+  and returns 0 if the relative index is infinite.
+
+# Main results
+
+- `card_mul_index` : `Nat.card H * H.index = Nat.card G`
+- `index_mul_card` : `H.index * Fintype.card H = Fintype.card G`
+- `index_dvd_card` : `H.index ∣ Fintype.card G`
+- `index_eq_mul_of_le` : If `H ≤ K`, then `H.index = K.index * (H.subgroupOf K).index`
+- `index_dvd_of_le` : If `H ≤ K`, then `K.index ∣ H.index`
+- `relindex_mul_relindex` : `relindex` is multiplicative in towers
+
+-/
+
+
+namespace Subgroup
+
+open BigOperators Cardinal
+
+variable {G : Type _} [Group G] (H K L : Subgroup G)
+
+/-- The index of a subgroup as a natural number, and returns 0 if the index is infinite. -/
+@[to_additive
+      "The index of a subgroup as a natural number,\nand returns 0 if the index is infinite."]
```

Comment:
This is an annoying bug in mathport. Please fix all the `\n` substrings into either line breaks or spaces.